drop-storage/src/lib.rs

```diff
@@ -32,4 +35,671 @@ impl Storage {
+            "INSERT INTO incoming_path_failed_states (path_id, status_code, bytes_received) \
+             VALUES ((SELECT id FROM incoming_paths WHERE transfer_id = ?2 AND path_id = ?2), ?3, \
+             ?4)",
+            transfer_id,
+            path_id,
+            error,
+            bytes_received
+        )
+        .execute(&mut conn)
+        .await
+        .map_err(error::Error::DBError)?;
+
+        Ok(())
+    }
+
+    pub async fn insert_outgoing_path_failed_state(
+        &self,
+        transfer_id: String,
+        path_id: String,
+        error: u32,
+        bytes_sent: i64,
+    ) -> Result<()> {
+        let mut conn = self.conn.acquire().await?;
+
+        sqlx::query!(
+            "INSERT INTO outgoing_path_failed_states (path_id, status_code, bytes_sent) VALUES \
+             ((SELECT id FROM outgoing_paths WHERE transfer_id = ?2 AND path_id = ?2), ?3, ?4)",
+            transfer_id,
+            path_id,
+            error,
+            bytes_sent
+        )
+        .execute(&mut conn)
+        .await
+        .map_err(error::Error::DBError)?;
+
+        Ok(())
+    }
+
+    pub async fn insert_outgoing_path_completed_state(
+        &self,
+        transfer_id: String,
+        path_id: String,
+    ) -> Result<()> {
+        let mut conn = self.conn.acquire().await?;
+
+        sqlx::query!(
+            "INSERT INTO outgoing_path_completed_states (path_id) VALUES ((SELECT id FROM \
+             outgoing_paths WHERE transfer_id = ?1 AND path_id = ?2))",
+            transfer_id,
+            path_id,
+        )
+        .execute(&mut conn)
+        .await
+        .map_err(error::Error::DBError)?;
+
+        Ok(())
+    }
+
+    pub async fn insert_incoming_path_completed_state(
+        &self,
+        transfer_id: String,
+        path_id: String,
+        final_path: String,
+    ) -> Result<()> {
+        let mut conn = self.conn.acquire().await?;
+
+        sqlx::query!(
+            "INSERT INTO incoming_path_completed_states (path_id, final_path) VALUES ((SELECT id \
+             from incoming_paths WHERE transfer_id = ?1 AND path_id = ?2), ?3)",
+            transfer_id,
+            path_id,
+            final_path
+        )
+        .execute(&mut conn)
+        .await
+        .map_err(error::Error::DBError)?;
+
+        Ok(())
+    }
+
+    pub async fn purge_transfers_until(&self, until_timestamp: i64) -> Result<()> {
+        let mut conn = self.conn.acquire().await?;
+
+        sqlx::query!(
+            "DELETE FROM transfers WHERE created_at < datetime(?1, 'unixepoch')",
+            until_timestamp
+        )
+        .execute(&mut conn)
+        .await
+        .map_err(error::Error::DBError)
+        .map(|_| ())
+    }
+
+    pub async fn purge_transfers(&self, transfer_ids: Vec<String>) -> Result<()> {
+        let mut conn = self.conn.acquire().await?;
+
+        let transfer_ids = transfer_ids.join(",");
+        sqlx::query!("DELETE FROM transfers WHERE id IN (?1)", transfer_ids)
```

Comment:
Is it possible to pass the array into the query directly?

Reply:
sadly sqlx currently doesn't support this